test_utils/src/cluster.rs

```diff
@@ -2,174 +2,514 @@
+        let mut primary_handlers = Node::spawn_primary(
+            key_pair,
+            self.committee.clone(),
+            &primary_store,
             self.parameters.clone(),
             /* consensus */ true,
             /* execution_state */ Arc::new(SimpleExecutionState),
             tx_transaction_confirmation,
             &registry,
         )
-        .await;
+        .await
+        .unwrap();
 
-        let node = Arc::new(NodeDetails {
-            id,
-            name,
-            registry,
-            handlers: h.unwrap(),
-            store_path,
+        let (tx, _) = tokio::sync::broadcast::channel(DEFAULT_CHANNEL_SIZE);
+        let transactions_sender = tx.clone();
+        // spawn a task to listen on the committed transactions
+        // and translate to a mpmc channel
+        let h = tokio::spawn(async move {
+            while let Some(t) = rx_transaction_confirmation.recv().await {
+                // send the transaction to the mpmc channel
+                transactions_sender
+                    .send(t)
+                    .expect("Couldn't send message to broadcast channel");
+            }
         });
 
-        // Insert to the nodes map
-        self.nodes.insert(id, node.clone());
+        // add the tasks's handle to the primary's handle so can be shutdown
+        // with the others.
+        primary_handlers.push(h);
 
-        Ok(node)
+        self.handlers.swap(Arc::new(primary_handlers));
+        self.store_path = store_path;
+        self.registry = registry;
+        self.tx_transaction_confirmation = tx;
     }
 
-    /// This method stops a node by the provided id. The method will return
-    /// either when the node has been successfully stopped or even when
-    /// the node doesn't exist.
-    pub fn stop_node(&mut self, id: usize) {
-        if let Some(node) = self.nodes.get_mut(&id) {
-            node.handlers.iter().for_each(|h| h.abort());
-            info!("Aborted node for id {id}");
+    pub fn stop(&self) {
+        self.handlers.load().iter().for_each(|h| h.abort());
+        info!("Aborted primary node for id {}", self.id);
+    }
+
+    /// This method returns whether the node is still running or not. We
+    /// iterate over all the handlers and check whether there is still any
+    /// that is not finished. If we find at least one, then we report the
+    /// node as still running.
+    fn is_running(&self) -> bool {
+        if self.handlers.load().is_empty() {
+            return false;
+        }
+
+        self.handlers.load().iter().any(|h| !h.is_finished())
+    }
+}
+
+#[derive(Clone)]
+pub struct WorkerNodeDetails {
+    pub id: WorkerId,
+    pub transactions_address: Multiaddr,
+    pub registry: Registry,
+    name: Ed25519PublicKey,
+    committee: SharedCommittee<Ed25519PublicKey>,
+    parameters: Parameters,
+    store_path: PathBuf,
+    handlers: Arc<ArcSwap<Vec<JoinHandle<()>>>>,
```

Comment:
This is probably a nit, I'll leave it to you to consider whether it's worth opening an issue to follow-up on this:

It is possible to me that the `ArcSwap` here might be an overkill: the `ArcSwap` is most useful to me when you have at least one mutation of a value from a non-empty state, i.e. `None` -> `Some(1)` -> `Some(42)`. (the `Option` is figurative here and just there to let me express the notion of "empty" state, not actually involved in this).

If I understand what you're trying to do correctly, you're more probably trying to initialize something from an empty state, and then bring it back to that empty state before setting it again, i.e. `None` -> `Some(1)`-> `None` -> `Some(42)`. For this consider that the `OnceCell` type has an initialization (and a `get_or_try_init` function) but also a `take` function that brings things back to this empty state.

Reply:
Ah thanks for the above. To be honest my intuition was also that probably the `ArcSwap` might be an overkill. Unfortunately I can't use `OnceCell` as `PrimaryNodeDetails` needs to be cloned (`#[derive(Clone)]` is on top) which ends up in linter errors. I've switched to using an `Rc<RefCell>` instead as I believe is what I really need (interior mutability behind a way to share across clones)